### PR TITLE
Backport modern section change logic from multires

### LINKED ISSUE #43: Resizing A Section The Player Isn't In Freezes The Game
TheXTech - Wut Bug

Wat... o.O

Okay, this is _weird_.

I'll try to provide as much information as I can...

After collecting all the stars, I saved and exited. Upon loading up, I appear at the door to the last level I exited. I re-enter... the music plays, the above GIF happens, and the game stops responding to any input.

There is no related terminal output with `--verbose` just the usual stuff.

The issue occurs on a specific save (Save 1), but not a mostly new one (Save 2).

I've attched my episode and saves below (minus custom graphics and music, they aren't needed for this or present in the GIF above, and the issue still ocured without them when I tested)

Again using the pre-built binary for Ubuntu 20.04 from [Release 1.3.4](https://github.com/Wohlstand/TheXTech/releases/tag/v.1.3.4)

[Stars of Night - Weird Bug.zip](https://github.com/Wohlstand/TheXTech/files/6018423/Stars.of.Night.-.Weird.Bug.zip)
I feel oddly talented at breaking the game XD

### PR DESCRIPTION
**Pull request by ds-sloth**
Partial port of the modern section change logic from multires. Does not contain the logic for `vScreen` position (`ScreenTop` / `ScreenLeft`) or for the "quick section change" used for star exits in the Yoshi's Archipelago episode.

This code is all well-tested and should be ready for immediate merge, but I'd like @Wohlstand's quick review before merging.

At 800x600, there will be two visible effects. First, bug #43 will not occur. Second, 2P camera merges / splits will be animated.

Fixes #43.

[section resize.lvl.zip](https://github.com/Wohlstand/TheXTech/files/12196216/section.resize.lvl.zip)

### CODE REVIEW

File: src/layers.cpp [hunk 1]
```diff
@@ -822,19 +840,132 @@ void ProcEvent(eventindex_t index, bool NoEffect)
...
+                                if(tX < 0)
+                                    xProp *= -1;
+                                if(tY < 0)
+                                    yProp *= -1;
+
+                                double maxShiftX = vScreen[1].Width / 4;
+                                double maxShiftY = vScreen[1].Height / 4;
+                                if(maxShiftX > 200)
+                                    maxShiftX = 200;
+                                if(maxShiftY > 150)
```
Comment (Wohlstand) [inline src/layers.cpp:919]: Empty line please between if conditions except `else if`.

File: src/layers.cpp [hunk 2]
```diff
@@ -822,19 +840,132 @@ void ProcEvent(eventindex_t index, bool NoEffect)
...
+
+                                // restrict to old level bounds
+                                if(-qScreenLoc[1].X < level[B].X)
+                                    qScreenLoc[1].X = -level[B].X;
+                                if(-qScreenLoc[1].X + vScreen[1].Width > level[B].Width)
+                                    qScreenLoc[1].X = -(level[B].Width - vScreen[1].Width);
+                                if(-qScreenLoc[1].Y < level[B].Y)
+                                    qScreenLoc[1].Y = -level[B].Y;
+                                if(-qScreenLoc[1].Y + vScreen[1].Height > level[B].Height)
+                                    qScreenLoc[1].Y = -(level[B].Height - vScreen[1].Height);
```
Comment (Wohlstand) [inline src/layers.cpp:933]: Here too:

> Empty line please between if conditions except `else if`.

File: src/layers.cpp [hunk 3]
```diff
@@ -851,14 +982,16 @@ void ProcEvent(eventindex_t index, bool NoEffect)
...
-                                qScreenLoc[1].Y = -level[Player[C].Section].Y;
-                            if(-qScreenLoc[1].Y + ScreenH /*FrmMain.ScaleHeight*/ > level[Player[C].Section].Height)
-                                qScreenLoc[1].Y = -(level[Player[C].Section].Height - ScreenH);
+                            // restrict to old level bounds
+                            if(-qScreenLoc[1].X < level[B].X)
+                                qScreenLoc[1].X = -level[B].X;
+                            if(-qScreenLoc[1].X + ScreenW /*FrmMain.ScaleWidth*/ > level[B].Width)
+                                qScreenLoc[1].X = -(level[B].Width - ScreenW);
+                            if(-qScreenLoc[1].Y < level[B].Y)
+                                qScreenLoc[1].Y = -level[B].Y;
```
Comment (Wohlstand) [inline src/layers.cpp:991]: And here:

> Empty line please between if conditions except `else if`.

File: src/layers.cpp [hunk 4]
```diff
@@ -822,19 +840,132 @@ void ProcEvent(eventindex_t index, bool NoEffect)
...
+                            {
+                                double dSquare = tX * tX + tY * tY;
+
+                                double xProp = tX * tX / dSquare;
+                                double yProp = tY * tY / dSquare;
+
+                                if(tX < 0)
+                                    xProp *= -1;
+                                if(tY < 0)
+                                    yProp *= -1;
```
Comment (Wohlstand) [inline src/layers.cpp:913]: And here too:

> Empty line please between if conditions except `else if`.

File: src/layers.cpp [hunk 5]
```diff
@@ -822,19 +840,132 @@ void ProcEvent(eventindex_t index, bool NoEffect)
...
+
+                            // set the qScreen!
+                            qScreen = true;
+                            qScreenLoc[1] = vScreen[1];
+
+                            // set the second qScreen if possible
+                            if(vScreen[2].Visible)
+                            {
+                                qScreenLoc[2] = vScreen[2];
+                            }
```
Comment (Wohlstand) [inline src/layers.cpp:900]: I guess, brackets here are pointless, just one single line will be fine.

**Review by Wohlstand — CHANGES_REQUESTED** [review]
Looks like fine, if not count code style quibbles. :)

**Comment by ds-sloth** [comment]
Fixed your style notes, thanks. Also added some additional comments. Note that I did add brackets around one single-line `else` block to balance with the multiline `if` block before it.

**Review by Wohlstand — APPROVED** [review]
I quickly checked the work of the thing, and seems works just fine!
Probably, if somebody also test the work of the thing, could give one another note. I'll merge this.

**Pull request merged**